### PR TITLE
feat: Adding a Loading state to buttons

### LINKED ISSUE #2625: [feat]: Adding a Loading state to buttons
# Feature request

## Please describe your feature

Currently, no limit is added to the buttons on all request interfaces. If the user has a poor network environment, there may be a situation where the create button is clicked multiple times in a row.

## Describe the solution you'd like

Add a global custom Hook to change the state of the button to Loading when clicking on a button in the Create, Submit, Delete etc. category and restore the state when the request interface is successful.

### PR DESCRIPTION
**Pull request by Silence-dream**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Add a global custom Hook to change the state of the button to Loading when clicking on a button in the Create, Submit, Delete etc. category and restore the state when the request interface is successful.

The specific scope of modification :
in routes submit,publish,offline button
in upstream submit and delete button
in service submit and delete button
in consumer submit and delete button
in proto submit and delete button
in plugin submit and delete button
in ssl submit and delete button

**Related issues**

fix/resolve #2625

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [x] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

**Comment by guoqqqi** [comment]
also @SkyeYoung PTAL

**Comment by Baoyuantop** [comment]
Hi @Silence-dream, excellent work. Can you get the CI passed? Feel free to ask if you need any help!

**Comment by Silence-dream** [comment]
> Hi @Silence-dream, excellent work. Can you get the CI passed? Feel free to ask if you need any help!

I have no experience with Cypress, can you give me some ideas?

**Comment by SkyeYoung** [comment]
@Silence-dream Just read the documentation in https://www.cypress.io/, which is very friendly. 😸

**Comment by guoqqqi** [comment]
> > Hi @Silence-dream, excellent work. Can you get the CI passed? Feel free to ask if you need any help!
> 
> I have no experience with Cypress, can you give me some ideas?

Hello, You can read some of the [older test cases](https://github.com/apache/apisix-dashboard/blob/master/web/cypress/e2e/route/batch-delete-route.cy.js) and if you have any questions about starting a test environment, you can refer to: 

https://github.com/apache/apisix-dashboard/blob/master/docs/en/latest/develop.md
https://github.com/apache/apisix-dashboard/blob/master/docs/en/latest/front-end-e2e.md

File: web/cypress/e2e/route/batch-delete-route.cy.js [hunk 1]
```diff
@@ -63,36 +63,38 @@ context('Create and Batch Deletion Routes', () => {
...
       // config label
       cy.contains('Manage').click();
-
+      // eslint-disable-next-line @typescript-eslint/no-loop-func
       cy.get(selector.drawerBody).within(($drawer) => {
         cy.wrap($drawer)
           .contains('button', 'Add')
           .should('not.be.disabled')
-          .click()
+          .click({ force: true })
```
Comment (LiteSun) [inline web/cypress/e2e/route/batch-delete-route.cy.js:77]: why does it need to add force here?

Comment (Silence-dream) [reply web/cypress/e2e/route/batch-delete-route.cy.js:77]: make click work

**Comment by Silence-dream** [comment]
image
image

image


Some errors I can't reproduce

**Comment by Silence-dream** [comment]
https://github.com/apache/apisix-dashboard/actions/runs/3209405460/jobs/5246820969
pluginTemplate-create-plugin-template-with-route.cy.js
My execution is normal.
image

**Comment by Silence-dream** [comment]
I don't understand why ci would fail, my execution was successful
image
image

In the previous implementation the results were successful
https://github.com/apache/apisix-dashboard/actions/runs/3209938717/jobs/5247221587
https://github.com/apache/apisix-dashboard/actions/runs/3210584308/jobs/5251519348
before
image
after
image

before
image
after
image


can you help me? @guoqqqi @LiteSun

File: web/src/hooks/useLatest.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (oil-oil) [inline web/src/hooks/useLatest.ts:1]: I think we can add `ahooks` to this library directly, without having to implement these hooks separately, what do you think? @guoqqqi @Baoyuantop @SkyeYoung

**Comment by Silence-dream** [comment]
Different results for the same commit, unbelievable

https://github.com/apache/apisix-dashboard/actions/runs/3213197938/jobs/5258560129
https://github.com/Silence-dream/apisix-dashboard/actions/runs/3213197898/jobs/5252694233

File: web/cypress/e2e/route/create-route-both-use-uri-uris.cy.js [hunk 3]
```diff
@@ -72,7 +72,7 @@ context('Create Route Both use uri and uris', () => {
     cy.contains('Route').click();
     cy.get(selector.empty).should('be.visible');
     cy.contains('Create').click();
-    cy.contains('Next').click().click();
```
Comment (guoqqqi) [inline web/cypress/e2e/route/create-route-both-use-uri-uris.cy.js:75]: The two clicks here are to ensure that the rendering of the page is finished, and this is a point that can be optimized. 
Please try to restore it first

File: web/cypress/e2e/route/create-route-with-advanced-matching-conditions.cy.js [hunk 4]
```diff
@@ -185,7 +189,9 @@ context('Create Route with advanced matching conditions', () => {
     cy.contains('Next').click();
     cy.get(selector.nodes_0_port).focus();
     cy.contains('Next').click();
+    cy.wait(timeout * 2);
```
Comment (guoqqqi) [inline web/cypress/e2e/route/create-route-with-advanced-matching-conditions.cy.js:192]: ditto

File: web/cypress/e2e/route/create-route-with-chash-upstream.cy.js [hunk 5]
```diff
@@ -102,6 +102,7 @@ context('Create and Edit Route With Custom CHash Key Upstream', () => {
     cy.get(selector.chash_key).should('value', data.custom_key);
     cy.get(selector.chash_key).clear().type(data.new_key);
     cy.contains('Next').click();
+    cy.wait(1000);
```
Comment (guoqqqi) [inline web/cypress/e2e/route/create-route-with-chash-upstream.cy.js:105]: Remove all new wait

File: web/cypress/e2e/route/create-route-with-advanced-matching-conditions.cy.js [hunk 6]
```diff
@@ -185,7 +189,9 @@ context('Create Route with advanced matching conditions', () => {
...
     cy.contains('Next').click();
+    cy.wait(timeout * 2);
```
Comment (guoqqqi) [inline web/cypress/e2e/route/create-route-with-advanced-matching-conditions.cy.js:194]: Remove all new wait

**Comment by guoqqqi** [comment]
Hi, @Silence-dream Please delete all new wait and resume the previous successive click attempts

File: web/src/hooks/useLatest.ts (hunk 2 above)
Comment (SkyeYoung) [reply web/src/hooks/useLatest.ts:1]: Yes, I think so. Just use the `ahooks` tested and used by many people. @guoqqqi @Silence-dream

**Comment by SkyeYoung** [comment]
> Hi, @Silence-dream Please delete all new wait and resume the previous successive click attempts
@Silence-dream May I ask why there are so many new `wait`s? Are there any other solutions?

**Comment by Silence-dream** [comment]
> > Hi, @Silence-dream Please delete all new wait and resume the previous successive click attempts
> > @Silence-dream May I ask why there are so many new `wait`s? Are there any other solutions?

The wait is added to ensure that the dom is obtained, but it works on my computer, both with and without the

File: web/src/hooks/useLatest.ts (hunk 2 above)
Comment (Silence-dream) [reply web/src/hooks/useLatest.ts:1]: > 
So I remove custom hooks and add ahooks?

Comment (Silence-dream) [reply web/src/hooks/useLatest.ts:1]: > Yes, I think so. Just use the `ahooks` tested and used by many people. @guoqqqi @Silence-dream



> Yes, I think so. Just use the `ahooks` tested and used by many people. @guoqqqi @Silence-dream

So I remove custom hooks and add ahooks?

Comment (Silence-dream) [reply web/src/hooks/useLatest.ts:1]: > Yes, I think so. Just use the `ahooks` tested and used by many people. @guoqqqi @Silence-dream

So I remove custom hooks and add ahooks?

**Comment by guoqqqi** [comment]
Hello @Silence-dream After the CI is run, you can download the video to help determine the cause of the bug.

**Comment by Silence-dream** [comment]
> Hello @Silence-dream After the CI is run, you can download the video to help determine the cause of the bug.

Thank you very much! It was very useful!

**Comment by guoqqqi** [comment]
NOTE: Since you are contributing to CI for the first time, you will need a Committer to trigger the CI to run.

File: web/src/hooks/useLatest.ts (hunk 2 above)
Comment (oil-oil) [reply web/src/hooks/useLatest.ts:1]: > > Yes, I think so. Just use the `ahooks` tested and used by many people. @guoqqqi @Silence-dream
> 
> So I remove custom hooks and add ahooks?

Yes.

File: web/cypress/e2e/route/create-route-with-search-service-and-set-priority.cy.js [hunk 7]
```diff
@@ -171,7 +172,7 @@ context('Create Route with search service name', () => {
     cy.contains(data.serviceName).siblings().contains('Delete').click();
     cy.contains('button', 'Confirm').click();
     cy.get(selector.notification).should('contain', data.deleteServiceSuccess);
-
+    cy.wait(4000);
```
Comment (guoqqqi) [inline web/cypress/e2e/route/create-route-with-search-service-and-set-priority.cy.js:175]: You can click here to remove the tip in the upper right corner, we should try not to waste time by waiting as much as possible:

```
  cy.get('.ant-notification-close-icon').click().should('not.exist');
```

File: web/src/pages/Consumer/List.tsx [hunk 8]
```diff
@@ -38,7 +38,7 @@ const Page: React.FC = () => {
   const [id, setId] = useState('');
   const [editorMode, setEditorMode] = useState<'create' | 'update'>('create');
   const { paginationConfig, savePageList, checkPageList } = usePagination();
-
+  const [deleteLoading, setDeleteLoading] = useState('');
```
Comment (guoqqqi) [inline web/src/pages/Consumer/List.tsx:41]: `deleteLoading` bad Variable name

File: web/src/pages/Plugin/List.tsx [hunk 9]
```diff
@@ -36,6 +36,7 @@ const Page: React.FC = () => {
   const [pluginList, setPluginList] = useState<PluginComponent.Meta[]>([]);
   const [name, setName] = useState('');
   const { paginationConfig, savePageList, checkPageList } = usePagination();
+  const [deleteLoading, setDeleteLoading] = useState('');
```
Comment (guoqqqi) [inline web/src/pages/Plugin/List.tsx:39]: ditto

File: web/src/pages/Route/Create.tsx [hunk 10]
```diff
@@ -32,6 +32,7 @@ import CreateStep4 from './components/CreateStep4';
 import { DEFAULT_STEP_1_DATA, DEFAULT_STEP_3_DATA } from './constants';
 import ResultView from './components/ResultView';
 import styles from './Create.less';
+import { useRequest } from 'ahooks';
```
Comment (guoqqqi) [inline web/src/pages/Route/Create.tsx:35]: move this line to 22

File: web/src/pages/Route/List.tsx [hunk 11]
```diff
@@ -59,6 +59,7 @@ import { DebugDrawView } from './components/DebugViews';
 import { RawDataEditor } from '@/components/RawDataEditor';
 import { EXPORT_FILE_MIME_TYPE_SUPPORTED } from './constants';
 import DataLoaderImport from '@/pages/Route/components/DataLoader/Import';
+import { useThrottleFn } from 'ahooks';
```
Comment (guoqqqi) [inline web/src/pages/Route/List.tsx:62]: move this line to 45

File: web/src/pages/Route/List.tsx [hunk 12]
```diff
@@ -108,18 +109,28 @@ const Page: React.FC = () => {
...
-        ? formatMessage({ id: 'page.route.publish' })
-        : formatMessage({ id: 'page.route.offline' });
-      handleTableActionSuccessResponse(
-        `${actionName} ${formatMessage({
-          id: 'menu.routes',
-        })} ${formatMessage({ id: 'component.status.success' })}`,
-      );
-    });
-  };
+  const [publishOfflineLoading, setPublishOfflineLoading] = useState<string>('');
```
Comment (guoqqqi) [inline web/src/pages/Route/List.tsx:112]: `loading` makes it read like a `boolean`

File: web/src/pages/Service/Create.tsx [hunk 13]
```diff
@@ -62,8 +62,9 @@ const Page: React.FC = (props) => {
       });
     }
   }, []);
-
+  const [submitLoading, setSubmitLoading] = useState(false);
```
Comment (guoqqqi) [inline web/src/pages/Service/Create.tsx:65]: Usually, we put the Hook at the top of the function scope

File: web/src/pages/Upstream/Create.tsx [hunk 14]
```diff
@@ -48,8 +48,10 @@ const Page: React.FC = (props) => {
       });
     }
   }, []);
+  const [submitLoading, setSubmitLoading] = useState(false);
```
Comment (guoqqqi) [inline web/src/pages/Upstream/Create.tsx:51]: ditto

**Review by SkyeYoung — CHANGES_REQUESTED** [review]

**Review by guoqqqi — APPROVED** [review]
LGTM, also cc @LiteSun @Baoyuantop PTAL

**Review by Baoyuantop — APPROVED** [review]

**Pull request merged**